toxcore/forwarding.c

```diff
@@ -0,0 +1,385 @@
+/* SPDX-License-Identifier: GPL-3.0-or-later
+ * Copyright © 2019-2022 The TokTok team.
+ */
+
+#include "forwarding.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "DHT.h"
+#include "ccompat.h"
+#include "timed_auth.h"
+
+struct Forwarding {
+    const Logger *log;
+    const Random *rng;
+    DHT *dht;
+    const Mono_Time *mono_time;
+    Networking_Core *net;
+
+    uint8_t hmac_key[CRYPTO_HMAC_KEY_SIZE];
+
+    forward_reply_cb *forward_reply_callback;
+    void *forward_reply_callback_object;
+
+    forwarded_request_cb *forwarded_request_callback;
+    void *forwarded_request_callback_object;
+
+    forwarded_response_cb *forwarded_response_callback;
+    void *forwarded_response_callback_object;
+};
+
+DHT *forwarding_get_dht(Forwarding *forwarding)
+{
+    return forwarding->dht;
+}
+
+#define SENDBACK_TIMEOUT 3600
+
+bool send_forward_request(Networking_Core *net, const IP_Port *forwarder,
+                          const uint8_t *chain_keys, uint16_t chain_length,
+                          const uint8_t *data, uint16_t data_length)
+{
+    const uint16_t len = forward_chain_packet_size(chain_length, data_length);
+    VLA(uint8_t, packet, len);
```

Comment:
This VLA can grow indefinitely given malicious inputs. I see `create_forward_chain_packet` does check its inputs, but `send_forward_request` doesn't.

Reply:
done